### PR TITLE
Feat: allow non-map tags

### LINKED ISSUE #54: Allow non-map tags
$ terratag -tags=var.default_tags
2020/09/24 23:09:46 Processing file cluster/ecs.tf
panic: invalid character 'v' looking for beginning of value

### PR DESCRIPTION
**Pull request by TomerHeber**
resolves #54.

Refactoring of testing.

### CODE REVIEW

**Comment by sagilaufer1992** [comment]
can you please add tests to this new tags format? I'm a big regex fan, but I don't trust anything regex related without some tests

**Review by sagilaufer1992 — COMMENTED** [review]
I'm not fully familiar with tagging rules, so I don't know wether this regex is sufficient or not. either way, should add tests to it, and make sure what's legal passes and what's not fails

File: main.go [hunk 1]
```diff
@@ -23,6 +23,8 @@ import (
 	"github.com/hashicorp/logutils"
 )
 
+var pairRegex = regexp.MustCompile(`^([a-zA-Z][\w-]*)=([\w-]+)$`)
```
[the remark this answers is not in the document]
Comment (TomerHeber) [reply main.go:26]: Based on terraform documentation, the answer is yes:
https://www.terraform.io/language/syntax/configuration#identifiers

**Comment by TomerHeber** [comment]
> I'm not fully familiar with tagging rules, so I don't know wether this regex is sufficient or not. either way, should add tests to it, and make sure what's legal passes and what's not fails

@sagilaufer1992 

The current testing infrastructure is limited. For example, there are no tests for valid vs invalid JSON input.
I could add tests for this use case but this will incur some (not minor) changes to the testing infrastructure.
Let me know if I should apply infrastructure changes as part of this PR in order to add the required tests.

**Comment by yaronya** [comment]
> Let me know if I should apply infrastructure changes as part of this PR in order to add the required tests.

@TomerHeber @sagilaufer1992 IMO I'd go for it 👍

**Comment by TomerHeber** [comment]
> > Let me know if I should apply infrastructure changes as part of this PR in order to add the required tests.
> 
> @TomerHeber @sagilaufer1992 IMO I'd go for it 👍

@yaronya  @sagilaufer1992 - sounds good. I will make changes after my refactoring PR is pushed.
I'll tag this as work in progress..

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -5,9 +5,34 @@ on:
     types: [opened, synchronize]
 
 jobs:
+  unit-tests:
```
Comment (TomerHeber) [inline .github/workflows/ci.yml:8]: New check for running unit tests.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -5,9 +5,34 @@ on:
...
+      - name: Go vet
+        run: |
+          ! go vet ./... | read
+      - name: Go staticcheck
+        uses: dominikh/staticcheck-action@v1.2.0
+        with:
+          version: "2022.1"
+          install-go: false
+      - name: Go Test
+        run: SKIP_INTEGRATION_TESTS=1 go test -v ./...
```
Comment (TomerHeber) [inline .github/workflows/ci.yml:31]: the env `SKIP_INTEGRATION_TESTS` will skip the integration test.
This allows us to safely run the command  `go test -v ./...` (which runs all the tests terratag).

File: internal/utils/utils.go [hunk 4]
```diff
@@ -4,7 +4,7 @@ import "sort"
 
 func SortObjectKeys(tagsMap map[string]string) []string {
 	var keys []string
-	for key, _ := range tagsMap {
+	for key := range tagsMap {
```
Comment (TomerHeber) [inline internal/utils/utils.go:7]: staticchecker warning.

File: terratag.go [hunk 5]
```diff
@@ -144,12 +152,20 @@ func tagFileResources(path string, dir string, filter string, tags string, tfVer
 	return perFileCounters
 }
 
-func jsonToHclMap(tags string) string {
+func toHclMap(tags string) (string, error) {
```
Comment (TomerHeber) [inline terratag.go:155]: returns an error instead of exiting (easier when unit testing).

File: terratag_test.go [hunk 6]
```diff
@@ -259,3 +268,37 @@ func filterSymlink(ss []string) (ret []string) {
 	}
 	return ret
 }
+
+func TestToHclMap(t *testing.T) {
```
Comment (TomerHeber) [inline terratag_test.go:272]: Added unit tests to cover some common use cases.

File: terratag_test.go [hunk 7]
```diff
@@ -259,3 +268,37 @@ func filterSymlink(ss []string) (ret []string) {
...
+	validCases := map[string]string{
+		`{"a":"b","c":"d"}`: `{"a"="b","c"="d"}`,
+		`a=b,c=d`:           `{"a"="b","c"="d"}`,
+		`a-key=b-value`:     `{"a-key"="b-value"}`,
+		"{}":                "{}",
+	}
+
+	for input, output := range validCases {
+		input, expectedOutput := input, output
```
Comment (sagilaufer1992) [inline terratag_test.go:281]:
```suggestion
	for input, expectedOutput := range validCases {
```

**Review by sagilaufer1992 — APPROVED** [review]
thanks for adding test

**Comment by TomerHeber** [comment]
> > > Let me know if I should apply infrastructure changes as part of this PR in order to add the required tests.
> > 
> > 
> > @TomerHeber @sagilaufer1992 IMO I'd go for it 👍
> 
> @yaronya @sagilaufer1992 - sounds good. I will make changes after my refactoring PR is pushed. I'll tag this as work in progress..

**Pull request merged**